File: src/freight_fate/discord_presence.py
"""Optional Discord Rich Presence: broad, privacy-safe "now playing" status.

This module is the *only* place that knows about Discord or the ``pypresence``
RPC library. Gameplay code reports a small :class:`PresenceState` (two short,
player-facing strings) and never touches the socket, the dependency, or the
throttle logic.

Everything here is best-effort and non-fatal by design. If Discord is closed,
unavailable, disconnected mid-session, or the optional ``pypresence`` package is
missing, the game must still start, play, and exit exactly as before -- so every
RPC interaction runs on a background thread and is wrapped so no error ever
propagates into the game loop.

Privacy: presence shows broad activity only (menu, terminal, driving, resting,
delivering) plus high-level route and cargo context. It never includes save
file paths, the driver's chosen name, internal debug data, or anything that is
not already visible game content.
"""

from __future__ import annotations

import asyncio
import contextlib
import logging
import os
import threading
import time
from collections.abc import Callable
from dataclasses import dataclass
from typing import Protocol

log = logging.getLogger(__name__)

# Discord application id for the "Freight Fate" rich-presence app. Register an
# application at https://discord.com/developers/applications and set its id here
# or via the FREIGHT_FATE_DISCORD_APP_ID environment variable. The placeholder
# below lets the integration run end to end; with an unregistered id Discord
# simply refuses the handshake and presence stays hidden (handled gracefully).
DEFAULT_CLIENT_ID = "1519334426453082162"

# Discord truncates the details/state lines at 128 characters.
MAX_FIELD_LEN = 128

# Discord rate-limits presence updates (~5 per 20s). Coalesce changes and never
# push more than one update per this many seconds; identical states are dropped.
MIN_UPDATE_INTERVAL_S = 15.0

# How often the worker re-evaluates while idle (also flushes a throttled change).
_WORKER_TICK_S = MIN_UPDATE_INTERVAL_S

# Backoff between connection attempts when Discord is not running.
_RECONNECT_INTERVAL_S = 30.0


@dataclass(frozen=True)
class PresenceState:
    """A broad, player-facing activity snapshot reported by gameplay code.

    ``activity`` is the headline line (e.g. "Driving a route"); ``detail`` is an
    optional secondary line (e.g. "Chicago to Dallas, steel coils"). Both are
    plain prose with no private data. Equality drives de-duplication, so two
    identical snapshots never trigger a redundant Discord update.
    """

    activity: str
    detail: str = ""


def _truncate(text: str, limit: int = MAX_FIELD_LEN) -> str:
    text = " ".join(text.split())  # collapse whitespace; keep it tidy
    if len(text) <= limit:
        return text
    return text[: max(0, limit - 1)].rstrip() + "…"  # ellipsis


def format_activity(state: PresenceState, *, start: float | None = None) -> dict:
    """Translate a :class:`PresenceState` into pypresence ``update`` kwargs.

    Pure and side-effect free so it is trivially testable. Empty fields are
    omitted so Discord does not render blank lines, and both visible lines are
    clamped to Discord's 128-character limit. ``start`` (an epoch timestamp)
    shows an elapsed-time counter for the whole session when provided.
    """
    payload: dict = {"details": _truncate(state.activity)}
    detail = _truncate(state.detail)
    if detail:
        payload["state"] = detail
    if start is not None:
        payload["start"] = int(start)
    return payload


def driving_presence(
    *,
    phase: str,
    origin: str,
    destination: str,
    cargo: str,
    fraction: float,
    moving: bool,
    truck_label: str = "",
) -> PresenceState:
    """Build a privacy-safe driving snapshot from broad gameplay values.

    Pure, so the route/cargo/progress wording is unit-testable without a live
    game. ``fraction`` is route progress in 0..1; it is rounded to the nearest
    five percent so a steadily advancing trip does not churn Discord updates.
    Only public game content (city names, cargo label, truck model) is used --
    never the driver name, save path, or internal state.
    """
    pct = max(0, min(100, round(max(0.0, min(1.0, fraction)) * 20) * 5))
    if phase == "pickup":
        activity = f"Deadheading to a pickup in {origin}" if origin else "Deadheading to a pickup"
        detail = f"Picking up {cargo}" if cargo else ""
        return PresenceState(activity, detail)
    verb = "Driving" if moving else "Stopped"
    if origin and destination:
        activity = f"{verb}: {origin} to {destination}"
    elif destination:
        activity = f"{verb} to {destination}"
    else:
        activity = f"{verb} a route"
    bits = []
    if cargo:
        bits.append(cargo)
    bits.append(f"{pct}% there")
    if truck_label:
        bits.append(truck_label)
    return PresenceState(activity, ", ".join(bits))


class _RpcClient(Protocol):
    """The slice of ``pypresence.Presence`` this module relies on."""

    def connect(self) -> object: ...
    def update(self, **kwargs: object) -> object: ...
    def clear(self) -> object: ...
    def close(self) -> object: ...


def pypresence_available() -> bool:
    """Whether the optional ``pypresence`` dependency can be imported."""
    try:
        import pypresence  # noqa: F401
    except Exception:
        return False
    return True


def _default_rpc_factory(client_id: str) -> _RpcClient:
    """Build a real pypresence client. Raises if the dependency is missing."""
    from pypresence import Presence

    return Presence(client_id)


RpcFactory = Callable[[str], _RpcClient]


def _teardown_rpc(rpc: object, *, send_close: bool) -> None:
    """Fully release a pypresence client, including its loop and pipe transport.

    ``pypresence``'s synchronous client builds a private asyncio event loop and
    (on Windows) a proactor pipe transport during the handshake, and does not
    always release them cleanly -- whether a connection is closed normally or the
    handshake itself fails (e.g. an unregistered app id rejected with InvalidID).
    Left dangling, the transport's finaliser later raises inside ``__del__`` and
    the loop leaks. To avoid that noisy, half-closed teardown we close the
    transport and then *pump the loop* so the disconnect actually completes
    before the loop is closed. Every step is best-effort and never raises.
    """
    loop = getattr(rpc, "loop", None)
    transport = getattr(rpc, "sock_writer", None)  # a proactor pipe transport
    if loop is None or loop.is_closed():
        # No asyncio loop to drain (a test double, or a never-connected client).
        if send_close:
            with contextlib.suppress(Exception):
                rpc.clear()  # type: ignore[attr-defined]
            with contextlib.suppress(Exception):
                rpc.close()  # type: ignore[attr-defined]
        return
    asyncio.set_event_loop(loop)
    try:
        if send_close:
            with contextlib.suppress(Exception):
                rpc.clear()  # type: ignore[attr-defined]
        if transport is not None:
            with contextlib.suppress(Exception):
                transport.close()
        # Let the proactor finish closing the pipe instead of yanking the loop.
        for _ in range(20):
            with contextlib.suppress(Exception):
                loop.run_until_complete(asyncio.sleep(0.01))
            if getattr(transport, "_sock", None) is None:
                break
        with contextlib.suppress(Exception):
            loop.run_until_complete(loop.shutdown_asyncgens())
    finally:
        with contextlib.suppress(Exception):
            loop.close()
        asyncio.set_event_loop(None)


class DiscordPresence:
    """Best-effort Discord Rich Presence service.

    Gameplay calls :meth:`update` with a :class:`PresenceState`; it returns
    immediately, never blocking the game loop. A daemon worker thread owns all
    socket I/O: it connects (retrying when Discord is closed), pushes the latest
    state subject to de-duplication and throttling, and reconnects if Discord
    goes away. :meth:`shutdown` clears the presence and joins the worker.

    The worker is optional (``threaded=False``) so tests can drive the exact
    same connect/throttle/send logic synchronously with an injected clock and a
    fake RPC client.
    """

    def __init__(
        self,
        *,
        enabled: bool = True,
        client_id: str | None = None,
        min_interval_s: float = MIN_UPDATE_INTERVAL_S,
        clock: Callable[[], float] = time.monotonic,
        rpc_factory: RpcFactory | None = _default_rpc_factory,
        session_start: float | None = None,
        threaded: bool = True,
    ) -> None:
        if client_id is None:
            client_id = os.environ.get("FREIGHT_FATE_DISCORD_APP_ID", DEFAULT_CLIENT_ID)
        # The feature is only live when it is enabled *and* there is a way to
        # talk to Discord. A missing dependency simply leaves it dormant.
        self._available = rpc_factory is not None
        self._enabled = bool(enabled) and self._available
        self._client_id = client_id
        self._min_interval = max(0.0, float(min_interval_s))
        self._clock = clock
        self._rpc_factory = rpc_factory
        self._session_start = session_start if session_start is not None else time.time()
        self._threaded = threaded

        self._lock = threading.Lock()
        self._desired: PresenceState | None = None
        self._last_sent: PresenceState | None = None
        self._last_send_t: float | None = None
        self._rpc: _RpcClient | None = None
        self._last_connect_attempt: float | None = None

        self._wake = threading.Event()
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._started = False

    # -- public API -----------------------------------------------------------

    @property
    def enabled(self) -> bool:
        return self._enabled

    @property
    def connected(self) -> bool:
        return self._rpc is not None

    def start(self) -> None:
        """Begin presence after app initialisation. Safe to call when disabled."""
        if not self._enabled or self._started:
            return
        self._started = True
        self._stop.clear()
        if self._threaded:
            self._thread = threading.Thread(target=self._run, name="discord-presence", daemon=True)
            self._thread.start()
        else:
            self._pump()

    def update(self, state: PresenceState | None) -> None:
        """Report the latest broad activity. Non-blocking and dedup-aware."""
        if not self._enabled or state is None:
            return
        with self._lock:
            if state == self._desired:
                return  # nothing changed; skip even the wakeup
            self._desired = state
        if self._threaded:
            self._wake.set()
        else:
            self._pump()

    def shutdown(self) -> None:
        """Clear the presence and stop the worker. Never raises."""
        self._stop.set()
        self._wake.set()
        thread = self._thread
        if thread is not None and thread.is_alive():
            thread.join(timeout=2.0)
        self._thread = None
        self._close()
        self._started = False

    def set_enabled(self, enabled: bool) -> None:
        """Toggle presence at runtime (e.g. from the settings menu)."""
        enabled = bool(enabled) and self._available
        if enabled == self._enabled:
            return
        self._enabled = enabled
        if enabled:
            # Re-show whatever the game last reported, reconnecting at once
            # rather than waiting out the idle backoff (this is a user action).
            with self._lock:
                self._last_sent = None
            self._last_connect_attempt = None
            self.start()
        else:
            was_started = self._started
            self._stop.set()
            self._wake.set()
            thread = self._thread
            if was_started and thread is not None and thread.is_alive():
                thread.join(timeout=2.0)
            self._thread = None
            self._close()
            self._started = False
            self._stop.clear()

    # -- worker / single-step logic ------------------------------------------

    def _run(self) -> None:
        while not self._stop.is_set():
            try:
                self._pump()
            except Exception:  # pragma: no cover - defensive belt-and-braces
                log.debug("Discord presence pump failed", exc_info=True)
            self._wake.wait(self._worker_wait())
            self._wake.clear()

    def _worker_wait(self) -> float:
        """Seconds to sleep before the next pump.

        Short when a reported change is still waiting out the throttle window,
        otherwise a lazy heartbeat. Capped so a throttled update always flushes.
        """
        with self._lock:
            pending = self._desired is not None and self._desired != self._last_sent
        if pending and self._last_send_t is not None:
            remaining = self._min_interval - (self._clock() - self._last_send_t)
            if remaining > 0:
                return min(remaining, _WORKER_TICK_S)
        return _WORKER_TICK_S

    def _pump(self) -> None:
        """One connect-then-maybe-send cycle. Swallows all RPC errors."""
        if self._stop.is_set():
            return
        if not self._ensure_connected():
            return
        with self._lock:
            desired = self._desired
            last_sent = self._last_sent
        if desired is None or desired == last_sent:
            return  # nothing new to show (de-dupe)
        now = self._clock()
        if self._last_send_t is not None and now - self._last_send_t < self._min_interval:
            return  # throttled; the worker re-checks after the window closes
        try:
            self._rpc.update(  # type: ignore[union-attr]
                **format_activity(desired, start=self._session_start)
            )
        except Exception:
            log.debug("Discord presence update failed; will reconnect", exc_info=True)
            self._close()
            return
        with self._lock:
            self._last_sent = desired
        self._last_send_t = now

    def _ensure_connected(self) -> bool:
        if self._rpc is not None:
            return True
        if self._rpc_factory is None:
            return False
        now = self._clock()
        if (
            self._last_connect_attempt is not None
            and now - self._last_connect_attempt < _RECONNECT_INTERVAL_S
        ):
            return False  # back off; Discord is probably not running
        self._last_connect_attempt = now
        rpc = None
        try:
            rpc = self._rpc_factory(self._client_id)
            rpc.connect()
        except Exception:
            log.debug("Discord not available; presence stays hidden", exc_info=True)
            # A failed handshake can still leave pypresence's event loop and
            # sockets open; tear them down so nothing leaks (e.g. a wrong app id
            # rejected with InvalidID).
            if rpc is not None:
                _teardown_rpc(rpc, send_close=False)
            return False
        self._rpc = rpc
        return True

    def _close(self) -> None:
        rpc = self._rpc
        self._rpc = None
        self._last_send_t = None
        with self._lock:
            self._last_sent = None
        if rpc is None:
            return
        _teardown_rpc(rpc, send_close=True)

File: crates/freight-fate/tests/discord_presence.rs
//! Port of `tests/test_discord_presence.py`: the optional Discord Rich
//! Presence service.
//!
//! These cover the behaviour that must hold regardless of whether Discord is
//! even installed: pure formatting, the disabled path, a missing/unavailable
//! RPC, de-duplication and throttling of updates, and clean shutdown. A fake
//! RPC client and an injected clock keep every test deterministic and free of
//! real sockets.

use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

use freight_fate::discord_presence::{
    driving_presence, format_activity, ActivityPayload, DiscordPresence, DiscordPresenceOptions,
    PresenceState, RpcClient, RpcFactory, DEFAULT_CLIENT_ID, MAX_FIELD_LEN,
};
use freight_fate::net::testing::ManualClock;

/// A stand-in for the RPC client recording every call.
#[derive(Default)]
struct FakeRpcState {
    connect_error: Option<String>,
    update_error: Option<String>,
    connects: usize,
    updates: Vec<ActivityPayload>,
    cleared: usize,
    closed: usize,
}

#[derive(Clone, Default)]
struct FakeRpc(Arc<Mutex<FakeRpcState>>);

impl FakeRpc {
    fn new() -> Self {
        Self::default()
    }

    fn with_connect_error(msg: &str) -> Self {
        let me = Self::new();
        me.0.lock().unwrap().connect_error = Some(msg.to_string());
        me
    }

    fn with_update_error(msg: &str) -> Self {
        let me = Self::new();
        me.0.lock().unwrap().update_error = Some(msg.to_string());
        me
    }

    fn connects(&self) -> usize {
        self.0.lock().unwrap().connects
    }

    fn updates(&self) -> Vec<ActivityPayload> {
        self.0.lock().unwrap().updates.clone()
    }

    fn cleared(&self) -> usize {
        self.0.lock().unwrap().cleared
    }

    fn closed(&self) -> usize {
        self.0.lock().unwrap().closed
    }

    fn factory(&self) -> RpcFactory {
        let me = self.clone();
        Arc::new(move |_cid: &str| Ok(Box::new(me.clone()) as Box<dyn RpcClient>))
    }
}

impl RpcClient for FakeRpc {
    fn connect(&mut self) -> Result<(), String> {
        let mut st = self.0.lock().unwrap();
        st.connects += 1;
        match &st.connect_error {
            Some(e) => Err(e.clone()),
            None => Ok(()),
        }
    }

    fn update(&mut self, payload: &ActivityPayload) -> Result<(), String> {
        let mut st = self.0.lock().unwrap();
        if let Some(e) = &st.update_error {
            return Err(e.clone());
        }
        st.updates.push(payload.clone());
        Ok(())
    }

    fn clear(&mut self) -> Result<(), String> {
        self.0.lock().unwrap().cleared += 1;
        Ok(())
    }

    fn close(&mut self) -> Result<(), String> {
        self.0.lock().unwrap().closed += 1;
        Ok(())
    }
}

/// A synchronous (non-threaded) service wired to a fake RPC and clock.
fn make_presence(
    rpc: &FakeRpc,
    clock: &Arc<ManualClock>,
    enabled: bool,
    min_interval_s: f64,
) -> DiscordPresence {
    DiscordPresence::new(DiscordPresenceOptions {
        enabled,
        client_id: Some("test-app-id".to_string()),
        min_interval_s,
        clock: clock.clock(),
        rpc_factory: Some(rpc.factory()),
        session_start: Some(1234.0),
        threaded: false,
    })
}

// -- application id -----------------------------------------------------------

#[test]
fn test_default_client_id_is_the_registered_freight_fate_app() {
    assert_eq!(DEFAULT_CLIENT_ID, "1519334426453082162");
    assert!(DEFAULT_CLIENT_ID.chars().all(|c| c.is_ascii_digit()));
}

#[test]
fn test_default_client_id_is_used_when_no_override_is_supplied() {
    std::env::remove_var("FREIGHT_FATE_DISCORD_APP_ID");
    let captured = Arc::new(Mutex::new(Vec::<String>::new()));
    let factory: RpcFactory = {
        let captured = Arc::clone(&captured);
        Arc::new(move |client_id: &str| {
            captured.lock().unwrap().push(client_id.to_string());
            Ok(Box::new(FakeRpc::new()) as Box<dyn RpcClient>)
        })
    };
    let presence = DiscordPresence::new(DiscordPresenceOptions {
        min_interval_s: 0.0,
        rpc_factory: Some(factory),
        threaded: false,
        ..DiscordPresenceOptions::default()
    });
    presence.start();
    presence.update(Some(PresenceState::activity("In the main menu")));
    presence.shutdown();

    assert_eq!(
        *captured.lock().unwrap(),
        vec![DEFAULT_CLIENT_ID.to_string()]
    );
}

// -- formatting ---------------------------------------------------------------

#[test]
fn test_format_activity_maps_fields_and_includes_start() {
    let payload = format_activity(
        &PresenceState::new("Driving a route", "Chicago to Dallas"),
        Some(1234.0),
    );
    assert_eq!(payload.details, "Driving a route");
    assert_eq!(payload.state.as_deref(), Some("Chicago to Dallas"));
    assert_eq!(payload.start, Some(1234));
}

#[test]
fn test_format_activity_omits_empty_state_line() {
    let payload = format_activity(&PresenceState::activity("In the main menu"), None);
    assert_eq!(payload.details, "In the main menu");
    assert!(payload.state.is_none());
    assert!(payload.start.is_none());
}

#[test]
fn test_format_activity_truncates_to_discord_limit() {
    let long_detail = "x".repeat(500);
    let payload = format_activity(&PresenceState::new(&"A".repeat(500), &long_detail), None);
    assert!(payload.details.chars().count() <= MAX_FIELD_LEN);
    assert!(payload.state.as_ref().unwrap().chars().count() <= MAX_FIELD_LEN);
    assert!(payload.details.ends_with('…'));
}

#[test]
fn test_driving_presence_is_privacy_safe_and_concise() {
    let state = driving_presence(
        "delivery",
        "Chicago",
        "Dallas",
        "steel coils",
        0.42,
        true,
        "Standard rig",
    );
    assert_eq!(state.activity, "Driving: Chicago to Dallas");
    assert!(state.detail.contains("steel coils"));
    assert!(state.detail.contains("40% there")); // rounded to nearest 5%
    assert!(state.detail.contains("Standard rig"));
    // Nothing private leaks into the strings.
    let blob = format!("{}{}", state.activity, state.detail).to_lowercase();
    assert!(!blob.contains("save") && !blob.contains('/') && !blob.contains('\\'));
}

#[test]
fn test_driving_presence_stopped_and_pickup_phrasing() {
    let stopped = driving_presence("delivery", "Reno", "Boise", "lumber", 0.9, false, "");
    assert!(stopped.activity.starts_with("Stopped"));

    let pickup = driving_presence("pickup", "Tampa", "Miami", "produce", 0.1, true, "");
    assert!(pickup.activity.to_lowercase().contains("pickup"));
    assert!(pickup.detail.contains("produce"));
}

#[test]
fn test_driving_presence_clamps_fraction() {
    assert!(driving_presence("delivery", "A", "B", "x", -1.0, true, "")
        .detail
        .contains("0% there"));
    assert!(driving_presence("delivery", "A", "B", "x", 2.0, true, "")
        .detail
        .contains("100% there"));
}

// -- disabled mode ------------------------------------------------------------

#[test]
fn test_disabled_never_touches_rpc() {
    let rpc = FakeRpc::new();
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, false, 15.0);
    presence.start();
    presence.update(Some(PresenceState::activity("In the main menu")));
    presence.shutdown();
    assert!(!presence.enabled());
    assert_eq!(rpc.connects(), 0);
    assert!(rpc.updates().is_empty());
}

// -- missing dependency / unavailable Discord ---------------------------------

#[test]
fn test_missing_dependency_leaves_service_dormant() {
    // rpc_factory=None models the RPC client being absent entirely.
    let presence = DiscordPresence::new(DiscordPresenceOptions {
        rpc_factory: None,
        threaded: false,
        ..DiscordPresenceOptions::default()
    });
    assert!(!presence.enabled());
    presence.start();
    presence.update(Some(PresenceState::activity("In the main menu")));
    presence.shutdown(); // must not panic
}

#[test]
fn test_discord_closed_is_handled_and_retried_after_backoff() {
    let rpc = FakeRpc::with_connect_error("Discord not running");
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, true, 15.0);
    presence.start();
    presence.update(Some(PresenceState::activity("In the main menu")));
    assert_eq!(rpc.connects(), 1); // tried once
    assert!(rpc.updates().is_empty()); // nothing sent; no crash
    assert!(!presence.connected());

    // Within the backoff window it does not hammer the socket.
    presence.update(Some(PresenceState::activity("At the terminal")));
    assert_eq!(rpc.connects(), 1);

    // After the backoff window it tries again.
    clock.advance(31.0);
    presence.update(Some(PresenceState::activity("Driving a route")));
    assert_eq!(rpc.connects(), 2);
    presence.shutdown();
}

#[test]
fn test_update_failure_drops_connection_for_reconnect() {
    let rpc = FakeRpc::with_update_error("pipe closed");
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, true, 15.0);
    presence.start();
    presence.update(Some(PresenceState::activity("Driving a route")));
    assert_eq!(rpc.connects(), 1);
    assert!(!presence.connected()); // disconnected after the failed update
    presence.shutdown(); // never panics
}

// -- throttling / de-duplication ---------------------------------------------

#[test]
fn test_identical_state_is_not_resent() {
    let rpc = FakeRpc::new();
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, true, 15.0);
    presence.start();
    presence.update(Some(PresenceState::new(
        "Driving a route",
        "Chicago to Dallas",
    )));
    presence.update(Some(PresenceState::new(
        "Driving a route",
        "Chicago to Dallas",
    )));
    presence.update(Some(PresenceState::new(
        "Driving a route",
        "Chicago to Dallas",
    )));
    assert_eq!(rpc.updates().len(), 1);
    presence.shutdown();
}

#[test]
fn test_changes_are_throttled_then_flushed() {
    let rpc = FakeRpc::new();
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, true, 15.0);
    presence.start();
    presence.update(Some(PresenceState::activity("In the main menu")));
    assert_eq!(rpc.updates().len(), 1);

    // A new state inside the throttle window is held back, not sent.
    clock.advance(5.0);
    presence.update(Some(PresenceState::activity("At the terminal")));
    assert_eq!(rpc.updates().len(), 1);

    // Once the window passes, the latest state flushes on the next report.
    clock.advance(11.0);
    presence.update(Some(PresenceState::activity("Driving a route")));
    assert_eq!(rpc.updates().len(), 2);
    assert_eq!(rpc.updates().last().unwrap().details, "Driving a route");
    presence.shutdown();
}

#[test]
fn test_first_update_sends_immediately() {
    let rpc = FakeRpc::new();
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, true, 15.0);
    presence.start();
    presence.update(Some(PresenceState::activity("In the main menu")));
    assert_eq!(rpc.updates().len(), 1);
    assert_eq!(rpc.updates()[0].details, "In the main menu");
    presence.shutdown();
}

// -- shutdown cleanup ---------------------------------------------------------

#[test]
fn test_shutdown_clears_and_closes_the_rpc() {
    let rpc = FakeRpc::new();
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, true, 15.0);
    presence.start();
    presence.update(Some(PresenceState::activity("Driving a route")));
    presence.shutdown();
    assert_eq!(rpc.cleared(), 1);
    assert_eq!(rpc.closed(), 1);
    assert!(!presence.connected());
}

#[test]
fn test_shutdown_is_idempotent_and_safe_without_start() {
    let presence = make_presence(&FakeRpc::new(), &ManualClock::new(), true, 15.0);
    presence.shutdown();
    presence.shutdown(); // no error on a second call or when never started
}

#[test]
fn test_set_enabled_toggles_runtime_state() {
    let rpc = FakeRpc::new();
    let clock = ManualClock::new();
    let presence = make_presence(&rpc, &clock, true, 15.0);
    presence.start();
    presence.update(Some(PresenceState::activity("Driving a route")));
    assert_eq!(rpc.updates().len(), 1);

    presence.set_enabled(false);
    assert!(!presence.enabled());
    assert_eq!(rpc.cleared(), 1); // disabling tears the presence down
    presence.update(Some(PresenceState::activity("At the terminal")));
    assert_eq!(rpc.updates().len(), 1); // ignored while disabled

    // Re-enabling reconnects and re-shows the last reported state at once.
    presence.set_enabled(true);
    assert_eq!(rpc.updates().len(), 2);
    assert_eq!(rpc.updates().last().unwrap().details, "Driving a route");

    // A fresh distinct state flushes after the throttle window.
    clock.advance(20.0);
    presence.update(Some(PresenceState::activity("At the terminal")));
    assert_eq!(rpc.updates().len(), 3);
    assert_eq!(rpc.updates().last().unwrap().details, "At the terminal");
    presence.shutdown();
}

// -- threaded path smoke ------------------------------------------------------

#[test]
fn test_threaded_service_sends_and_shuts_down() {
    let rpc = FakeRpc::new();
    let presence = DiscordPresence::new(DiscordPresenceOptions {
        client_id: Some("test-app-id".to_string()),
        min_interval_s: 0.0,
        rpc_factory: Some(rpc.factory()),
        threaded: true,
        ..DiscordPresenceOptions::default()
    });
    presence.start();
    presence.update(Some(PresenceState::activity("In the main menu")));
    // Give the worker a moment to connect and push the first update.
    let deadline = Instant::now() + Duration::from_secs(2);
    while rpc.updates().is_empty() && Instant::now() < deadline {
        thread::sleep(Duration::from_millis(10));
    }
    presence.shutdown();
    assert!(rpc.connects() >= 1);
    assert!(!rpc.updates().is_empty());
    assert_eq!(rpc.closed(), 1);
}
